README.md
Dirvish
=======

Dirvish is an rsync based backup script.

This repository is a publicly accessible backup of my old private SVN branch.

Please note that this code is largely untested and SHOULD NOT be used by anyone.

Pleade DO NOT open pull requests or issues on this repository. It is unmaintained.


debian/changelog
dirvish (1.3.2-r657-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Sun, 08 Feb 2009 00:41:43 +0100

dirvish (1.3.2-r652-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 04 Feb 2009 19:34:46 +0100

dirvish (1.3.2-r634-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 28 Jan 2009 17:46:52 +0100

dirvish (1.3.2-r632-1) unstable; urgency=low

  * Initial release

 -- Dominik Schulz <dominik.schulz@gauner.org>  Tue, 27 Jan 2009 16:18:14 +0100


dirvish-expire.pl
#!/usr/bin/perl
# dirvish-expire
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 660 $
# Revision date   : $Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-expire.pl $

#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################
#
#----------------------------------------------------------------------------
# Revision information
#----------------------------------------------------------------------------

my %CodeID = (
    Rev    => '$Rev: 660 $'     ,
    Date   => '$Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-expire.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
no strict qw(refs);
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

use File::Spec::Functions;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
my $Options = reset_options(\&usage, @ARGV);     # initialize the %$Options hash
load_master_config('f', $Options);

GetOptions($Options, qw(
    quiet!
    vault=s
    time=s
    pidfile=s
    tree!
    no-run|dry-run
    version
    help|?
    )) or &usage();

ref($$Options{vault}) and $$Options{vault} = undef;

my $expire_time = undef;
$$Options{time} and $expire_time = parsedate($$Options{time});
$expire_time ||= time;

my $pidfile = $$Options{pidfile};
$pidfile =~ s/\/dirvish[.]pid/\/dirvish-expire.pid/;

if(!check_pidfile($pidfile)) {
	print "$0 already running. Aborting at ".strftime('%H:%M:%S', localtime)." - PID: ".$$."\n";
	exit 255; # Abort
}

my @expires = ();
my %unexpired = ();

if ($$Options{vault})
{
    my $b;
    my $bank;
    for $b (@{$$Options{bank}})
    {
       	if (-d "$b/$$Options{vault}")
        {
            $bank = $b;
            last;
        }
    }
    $bank or seppuku 252, "Cannot find vault $$Options{vault}";
    Dirvish::find(catdir($bank, $$Options{vault}), $expire_time, $Options, \@expires, \%unexpired);
} else {
    for my $bank (@{$$Options{bank}})
    {
        Dirvish::find(catdir($bank), $expire_time, $Options, \@expires, \%unexpired);
    }
}

scalar(@expires) or exit 0;

if (!$$Options{quiet})
{
    printf "Expiring images as of %s\n",
        strftime('%Y-%m-%d %H:%M:%S', localtime($expire_time));
    $$Options{vault} and printf "Restricted to vault %s\n",
        $$Options{vault};
    print "\n";

    printf "%-15s %-15s %-16.16s  %s\n",
        qw(VAULT:BRANCH IMAGE CREATED EXPIRED);
}

# add nice and ionice to rm execution
my $basecmd = (($$Options{nice} && $$Options{nice} > -1 && &check_nice()) ? &check_nice()." -".$$Options{nice} : '');
$basecmd .= (($$Options{ionice} && $$Options{ionice} > -1 && &check_ionice()) ? &check_ionice()." -n".$$Options{ionice} : '');

# Iterate over the expired images and remove them
for my $expire (sort imsort @expires)
{
    my ($created, $expired);
    ($created = $$expire{created}) =~ s/:\d\d$//;
    ($expired = $$expire{expire}) =~ s/:\d\d$//;
    
    if (!$unexpired{$$expire{vault}}{$$expire{branch}})
    {
        printf "cannot expire %s:%s:%s No unexpired good images\n",
            $$expire{vault},
            $$expire{branch},
            $$expire{image};
        $$expire{status} =~ /^success/
            and ++$unexpired{$$expire{vault}}{$$expire{branch}};

        # By virtue of the sort order this will be the newest 
        # image so that older ones can be expired.
        next;
    }
    $$Options{quiet} or printf "%-15s %-15s %-16.16s  %s\n",
        $$expire{vault} . ':' .  $$expire{branch},
        $$expire{image},
        $created,
        $expired;

	print($basecmd." rm -rf $$expire{path}/tree\n") if($$Options{'no-run'} && !$$Options{quiet});
    system($basecmd." rm -rf $$expire{path}/tree") unless $$Options{'no-run'};
    
    $$Options{tree} and next;

	print($basecmd." rm -rf $$expire{path}\n") if($$Options{'no-run'} && !$$Options{quiet});
    system($basecmd." rm -rf $$expire{path}") unless $$Options{'no-run'};

}
exit 0;
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
# Sort images
sub imsort {
	## WARNING:  don't mess with the sort order, it is needed so that if
	## WARNING:  all images are expired the newest will be retained.
	$$a{vault} cmp $$b{vault}
	  || $$a{branch} cmp $$b{branch}
	  || $$a{created} cmp $$b{created};
}
# Show usage information
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    print STDERR <<EOUSAGE;
USAGE
	dirvish.expire OPTIONS
	
OPTIONS
	--time date_expression
	--[no]tree
	--vault vault_name
	--no-run
	--quiet
EOUSAGE

	exit 255;
}
# Remove the pidfile on exit
END {
	# remove the pidfile on exit
	remove_pidfile($pidfile);
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------


dirvish-runall.pl
#!/usr/bin/perl
# dirvish-runall
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 652 $
# Revision date   : $Date: 2009-02-04 19:34:29 +0100 (Mi, 04 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-runall.pl $

#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################

#----------------------------------------------------------------------------
# Revision information
my %CodeID = (
    Rev    => '$Rev: 652 $'     ,
    Date   => '$Date: 2009-02-04 19:34:29 +0100 (Mi, 04 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-runall.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
our $VERSION;

my $Options = {
        version         => sub {
                        print STDERR "dirvish version $VERSION\n";
                        exit(0);
                },
        help            => \&usage,
};

$Options = reset_options( \&usage, @ARGV);	# initialize the %$Options hash
load_master_config(undef, $Options);		# load master config into $Options

GetOptions($Options, qw(
        config=s
        pidfile=s
        quiet no-run|dry-run
        version help
    )) or &usage();

$$Options{Dirvish} ||= 'dirvish';

$$Options{'no-run'} and $$Options{quiet} = 0;

my $errors = 0;

if(!check_pidfile($$Options{pidfile})) {
	print "$0 already running. Aborting at ".strftime('%H:%M:%S', localtime)." - PID: ".$$."\n";
	exit 255; # Abort
}

my $thmutex = undef;						# Thread mutex
my @threads = undef;						# List of running threads
if($$Options{'Threads'}) {
	use Thread qw(async);
	use Thread::Semaphore;
	$thmutex = Thread::Semaphore->new($$Options{'Threads'});	# Limit the number of running threads
	@threads = ();
}

for my $sched (@{$$Options{Runall}})
{
	my ($vault, $itime) = split(/\s+/, $sched);
	print "runall - vault: $vault\n";
	my $cmd = "$$Options{Dirvish} --vault $vault";
	$itime and $cmd .= qq[ --image-time "$itime"];
	$$Options{quiet}
		or printf "%s %s\n", strftime('%H:%M:%S', localtime), $cmd;
	if($$Options{'Threads'}) {
		my $t = async {
			$thmutex->down();			# Semaphore will block until we have a free slot
			$$Options{'no-run'} and $cmd .= " --no-run";	# Actually execute dirvish, even when running
			# in no-run mode, but make it also run in no-run mode
			print "Thread ".Thread->self->tid()." running cmd: $cmd\n" unless $$Options{quiet};
		    my $status = system($cmd);
		    $thmutex->up();				# Release semaphore lock so another thread can run
	    	return $status;
	    };
	    push(@threads,$t);
    } else {
    	# Without threads, everything is done sequential
    	my $status = system($cmd) unless $$Options{'no-run'};
	    $status < 0 || $status / 256 and ++$errors;
    }
}

if($$Options{'Threads'}) {
	# Wait for the threads to finish
	for my $th (@threads) {
		my $status = $th->join();
		$status < 0 || $status / 256 and ++$errors;
	}
}

remove_pidfile($$Options{pidfile});

$$Options{quiet}
    or printf "%s %s\n", strftime('%H:%M:%S', localtime), 'done';


exit ($errors < 200 ? $errors : 199);
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    $! and exit(255); # because getopt seems to send us here for death

    print STDERR <<EOUSAGE;
USAGE
	dirvish-runall OPTIONS
OPTIONS
	--config configfile
	--no-run
	--quiet
	--version
EOUSAGE

	exit 255;
}
# Remove the pidfile on exit
END {
	# remove the pidfile on exit
	remove_pidfile($$Options{pidfile});
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------


wif/dirvish-webif.pl
#!/usr/bin/perl -w
# dirvish
# 1.3.X series
# Copyright 2009 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 654 $
# Revision date   : $Date: 2009-02-05 22:10:21 +0100 (Do, 05 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/wif/dirvish-webif.pl $
#
#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################
#
# This is a web-based configuration and control frontend for dirvish.
# http://snippets.dzone.com/posts/show/257

# Includes
use lib "../.";

use warnings;
use strict;
use HTTP::Daemon;
use HTTP::Status;
use Dirvish;
use Data::Dumper;
use Thread qw(async);

# Initialisation
my $template = `cat dirvish.html`;
# initialize the %$Options hash
my $Options = reset_options( \&usage, @ARGV);
load_master_config('', $Options); 

#dump_config($Options);

print "Port: $ARGV[0]\n";

# Setup Webserver
my $d = HTTP::Daemon->new(
	LocalAddr	=> 'localhost',
	LocalPort	=> $ARGV[0],
) || die;

print "Please contact me at: <URL:", $d->url, ">\n";
CONNECTION: while (my $c = $d->accept) {
	my $t = async {
		REQUEST: while (my $r = $c->get_request) {
			#print "Thread: ".Thread->self->tid()." - Request: ".$r->url." - Path: - ".$r->url->path."\n";
			my $resp = get_response($c,$r);
        	$c->send_response($resp) if $resp;
    	}
    	$c->close;
    	undef($c);
	};
}

END {
	$d->shutdown();
}

# Subs
sub get_response {
	my $c = shift; # connection
	my $r = shift; # request
	my $body = $template;
	# Top-Navigation
	my $topnav = '<li><a href="/browse/">Browse</a></li>';
	$topnav .= '<li><a href="/search">Search</a></li>';
	$topnav .= '<li><a href="/settings">Settings</a></li>';
	$topnav .= "<li>Status: ";
	my $expire_pidfile = $$Options{'pidfile'};
	$expire_pidfile =~ s/\/dirvish[.]pid/\/dirvish-expire.pid/;
	if(-f $$Options{'pidfile'} || -f $expire_pidfile) {
		my $pid = `cat $$Options{'pidfile'}`;
		chomp($pid);
		# TODO check if it is really running, see check_pidfile
		$topnav .= 'Running - <a href="/stop_backup">Stop Backup</a>';
	} else {
		$topnav .= 'Idle - <a href="/start_backup">Start Backup</a>';
	}
	$topnav .= "</li>\n";
	# Left-Navigation
	my $left_title = "Browse";
	my $left_nav;
	foreach my $vault (@{$$Options{'Runall'}}) {
		$left_nav .= '<li>Vault <a href="/browse/'.$vault.'/">'.$vault.'</a></li>';
	}
	# Content Default
	my $content_title = "Welcome to Dirivsh";
	my $content_body = "&nbsp;";
	# Content
	my $path = $r->url->path;
	$path = url_decode($path);
	$path =~ s/[.][.]//g; # remove double dots
	$path =~ s/(\/\s*\/)+/\//g; # remove repeated-slashes
	my @path = split("/",$path);
	if($path =~ m/^\/settings/) {
		# TODO handle save settings
		$content_title = "Settings";
		$content_body .= '<form action="/settings" method="POST">';
		$content_body .= '<input type="hidden" name="save" value="1" /><br />';
		$content_body .= 'Bank: <input type="text" value="'.$$Options{'bank'}[0].'" name="bank" /><br />'; # TODO muliple banks
		$content_body .= 'Index-Type: <input type="text" value="'.$$Options{'index'}.'" name="index" /><br />'; # TODO select (none,gzip,bzip2)
		$content_body .= 'Ionice: <input type="text" value="'.$$Options{'ionice'}.'" name="ionice" /><br />'; # TODO select (0-7)
		$content_body .= 'Nice: <input type="text" value="'.$$Options{'nice'}.'" name="nice" /><br />'; # TODO select (0-19)
		$content_body .= 'Rsh: <input type="text" value="'.$$Options{'rsh'}.'" name="rsh" /><br />';
		$content_body .= '<input type="submit" name="Submit" value="Submit" />';
		$content_body .= '</form>';
	} elsif($path =~ m/^\/browse/) {
		# show file browser
		shift(@path); # (empty) root
		shift(@path); # action
		my $vault = shift(@path); # vault
		if(!$vault) {
			$content_title = "Browsing Vaults";
			$content_body = "";
			# show list of vaults and return
			foreach my $vault (@{$$Options{'Runall'}}) {
				$content_body .= 'Vault <a href="/browse/'.$vault.'/">'.$vault.'</a><br />';
			}
			goto BODY;
		}
		# TODO find correct bank
    	my $bank = "";
	    foreach my $bankc (@{$$Options{bank}})
    	{
        	if (-d "$bankc/$vault")
        	{
        		$bank = $bankc;
	            last;
        	}
    	}
    	if(!$bank) {
    		$content_body = "Error. Vault not found in Banks.";
    		goto BODY;
    	}
		my $image = shift(@path); # image
		# show a list of image in the navigation
		$left_nav = "";
		foreach my $fvault (@{$$Options{'Runall'}}) {
			$left_nav .= '<li>Vault <a href="/browse/'.$fvault.'/">'.$fvault.'</a></li>';
			if($fvault eq $vault) {
				# show image in reverse chronological order
				foreach my $image (reverse sort glob("$bank/$vault/*")) {
					my @image_path = split("/", $image);
					if($image_path[-1] =~ m/^\d{4}/) {
						$left_nav .= 'Image <a href="/browse/'.$vault.'/'.$image_path[-1].'/">'.$image_path[-1].'</a>';
						$left_nav .= "<br />\n";
					}
				}
			}
		}
		if(!$image) {
			$content_title = "Browsing Vault $vault";
			$content_body = '<a href="/browse/">Up ..</a><br />';
			# show list of images and return
			# show image in reverse chronological order
			foreach my $image (reverse sort glob("$bank/$vault/*")) {
				my @image_path = split("/", $image);
				if($image_path[-1] =~ m/^\d{4}/) {
					$content_body .= 'Image <a href="/browse/'.$vault.'/'.$image_path[-1].'/">'.$image_path[-1].'</a>';
					$content_body .= ' - <a href="/log/'.$vault.'/'.$image_path[-1].'">Log</a> - <a href="/summary/'.$vault.'/'.$image_path[-1].'">Summary</a>';
					$content_body .= "<br />\n";
				}
			}
			goto BODY;
 		}
		$path = join("/",@path);
		#$path =~ s/[.][.]//g; # remove double dots
		#$path =~ s/(\/\s*\/)+/\//g; # remove repeated-slashes
		#print "New-Path: $path\n";
		# Get directory contents
		#my $fspath = $$Options{bank}[0]."/".$vault."/".$image."/tree/".$path;
		my $fspath = $bank."/".$vault."/".$image."/tree/".$path;
		@path = split("/",$path);
		#pop(@path);
		#print "Path1: @path\n";
		if(-d $fspath) {
			# show content of dir
			unshift(@path, $image);
			unshift(@path, $vault);
			$content_title = "Browsing ".join("/",@path)."/";
			pop(@path); # move one dir up for ..
			$content_body = '<a href="/browse/'.join("/",@path).'/">Up ..</a><br />';
			foreach my $file (glob("$fspath/*")) {
				my @file = split("/",$file);
				if(-f $file) {
					$content_body .= '<a href="'.url_encode($file[-1]).'"><img src="/page_white.png" border="0" /> '.$file[-1].' '.
					sprintf("%.2f",((stat($file))[7])/1024)
					.'kb </a><br />';
				} elsif(-d $file) {
					$content_body .= '<a href="'.url_encode($file[-1]).'/"><img src="/folder.png" border="0" /> '.$file[-1].'</a><br />';
				} else {
					$content_body .= '<a href="'.url_encode($file[-1]).'">??? '.$file[-1].'</a><br />';
				}
			}
		} elsif(-f $fspath) {
			#print "Sending File: $fspath\n";
			# send file
			$c->send_file_response($fspath);
			return 0;
		} else {
			#print "Error 404\n";
			# show error 404
			$content_body = "Error 404<br />\n";
		}
	} elsif($path =~ m#^/start_backup#) {
		# TODO start backup
		print "Would execute /etc/dirvish/dirvish-cronjob\n";
		$content_title = "Starting Backup";
		$content_body = "Starting Backup ...<br />";
	} elsif($path =~ m#^/stop_backup#) {
		# TODO stop backup
		print "Would execute kill `cat /var/run/dirvish.pid`\n";
		$content_title = "Stopping Backup";
		$content_body = "Stopping Backup ...<br />";
	} elsif($path =~ m#^/search#) {
		# TODO show search form
		$content_title = "Search";
		$content_body = '<form action="/search" method="POST">';
		$content_body .= '<input type="hidden" name="search" value="1" /><br />';
		$content_body .= 'Bank: <input type="text" name="query" /><br />';
		$content_body .= '<input type="submit" name="Submit" value="Submit" />';
		$content_body .= '</form>';
	} elsif($path =~ m#^/log#) {
		shift(@path); # (empty) root
		shift(@path); # action
		my $vault = shift(@path); # vault
		my $image = shift(@path); # image
		$content_title = "Log for $vault/$image";
		my $logpath = $$Options{bank}[0]."/".$vault."/".$image."/log";
		if(-f $logpath) {
			$content_body .= "<pre>\n";
			$content_body .= `cat $logpath`;
			$content_body .= "</pre>\n";
		}
	} elsif($path =~ m#^/summary#) {
		shift(@path); # (empty) root
		shift(@path); # action
		my $vault = shift(@path); # vault
		my $image = shift(@path); # image
		$content_title = "Summary for $vault/$image";
		my $logpath = $$Options{bank}[0]."/".$vault."/".$image."/summary";
		if(-f $logpath) {
			$content_body .= "<pre>\n";
			$content_body .= `cat $logpath`;
			$content_body .= "</pre>\n";
		}
	} elsif($path =~ m/\.png$/ && -e ".$path") {
		$c->send_file_response(".$path");
		return 0;
	}
	BODY:
	$body =~ s/_TOP_NAV_/$topnav/ if $topnav;
	$body =~ s/_LEFT_TITLE_/$left_title/ if $left_title;
	$body =~ s/_LEFT_NAV_/$left_nav/ if $left_nav;
	$body =~ s/_CONTENT_TITLE_/$content_title/ if $content_title;
	$body =~ s/_TITLE_/$content_title/ if $content_title;
	$body =~ s/_CONTENT_BODY_/$content_body/ if $content_body;
	# TODO replace 
	return res($body);
}

sub res {
	HTTP::Response->new(
		RC_OK, OK => [ 'Content-Type' => 'text/html' ], shift
	)
}
# http://support.internetconnection.net/CODE_LIBRARY/Perl_URL_Encode_and_Decode.shtml
sub url_encode {
	my $str = shift;
	$str =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;
	return $str;
}
# http://support.internetconnection.net/CODE_LIBRARY/Perl_URL_Encode_and_Decode.shtml
sub url_decode {
	my $str = shift;
	$str =~ s/\%([A-Fa-f0-9]{2})/pack('C', hex($1))/seg;
	return $str;
}
# EOF

dirvish-locate.pl
#!/usr/bin/perl
# dirvish-locate
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 660 $
# Revision date   : $Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-locate.pl $
#
#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################
#
#----------------------------------------------------------------------------
# Revision information
#----------------------------------------------------------------------------
my %CodeID = (
    Rev    => '$Rev: 660 $'     ,
    Date   => '$Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-locate.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

use File::Spec::Functions;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
my $KILLCOUNT = 1000;
my $MAXCOUNT = 100;

my $Options = reset_options( \&usage, @ARGV);	# initialize the %$Options hash
load_master_config('f', $Options);				# load master config into $Options

GetOptions($Options, qw(
    version
    help|?
    )) or &usage();

my $Vault = shift;
my $Branch = undef;
$Vault =~ /:/ and ($Vault, $Branch) = split(/:/, $Vault);
my $Pattern = shift;

$Vault && length($Pattern) or &usage();
# prepend dot if asterisk or question mark is
# the first character. Make rsync-like patterns like *.xml work
$Pattern = ".".$Pattern if($Pattern =~ m/^(\*|\?)/);

my $fullpattern = $Pattern;
my $partpattern = undef;
$fullpattern =~ /\$$/ or $fullpattern .= '[^/]*$';
($partpattern = $fullpattern) =~ s/^\^//;

my $bank = undef;
for $b (@{$$Options{bank}})
{
    -d catdir($b,$Vault) and $bank = $b;
}
$bank or seppuku 220, "No such vault: $Vault";

opendir VAULT, catdir($bank,$Vault) or seppuku 221, "cannot open vault: $Vault";
my @invault = readdir(VAULT);
closedir VAULT;

my @images = ();
for my $image (@invault)
{
    $image eq 'dirvish' and next;
    my $imdir = catdir($bank,$Vault,$image);
    -f catfile($imdir,"summary") or next;
    (-l $imdir && $imdir =~ /current/) and next; # skip current-symlink
    my $conf = loadconfig('R', catfile($imdir,"summary"), $Options) or next;
    $$conf{Status} eq 'success' || $$conf{Status} =~ /^warn/
        or next;
    $$conf{'Backup-complete'} or next;
    $Branch && $$conf{branch} ne $Branch and next;
	
    unshift @images, {
        imdir   => $imdir,
        image   => $$conf{Image},
        branch  => $$conf{branch},
        created => $$conf{'Backup-complete'},
    }
}

my $imagecount = 0;
my $pathcount = 0;
my $path = undef;
my %match = ();
for my $image (sort(imsort_locate @images))
{
    my $imdir = $$image{imdir};

    my $index = undef;
    -f catfile($imdir,"index.bz2") and $index = "bzip2 -d -c ".catfile($imdir,"index.bz2")."|";
    -f catfile($imdir,"index.gz") and $index = "gzip -d -c ".catfile($imdir,"index.gz")."|";
    -f catfile($imdir,"index") and $index = "<".catfile($imdir,"index");
    $index or next;

    ++$imagecount;

	# can't use three-fold open here, see above
    open(INDEX, $index) or next;
    while (<INDEX>)
    {
        chomp;

        m($partpattern) or next;
		# this parse operation is too slow.  It might be faster as a
		# split with trimmed leading whitespace and remerged modtime
        my $f = { image => $image };
        (
             $$f{inode},
            $$f{blocks},
            $$f{perms},
            $$f{links},
            $$f{owner},
            $$f{group},
            $$f{bytes},
            $$f{mtime},
            $path
        ) = m<^
            \s*(\S+)        # inode
            \s+(\S+)        # block count
            \s+(\S+)        # perms
            \s+(\S+)        # link count
            \s+(\S+)        # owner
            \s+(\S+)        # group
            \s+(\S+)        # byte count
            \s+(\S+\s+\S+\s+\S+)    # date
            \s+(\S.*)        # path
        $>x;
        $$f{perms} =~ /^[dl]/ and next;
        $path =~ m($fullpattern) or next;

        exists($match{$path}) or ++$pathcount;
        push @{$match{$path}}, $f;
    }
    if ($pathcount >= $KILLCOUNT)
    {
        print "dirvish-locate: too many paths match pattern, interrupting search\n";
        last;
    }
}

printf "%d matches in %d images\n", $pathcount, $imagecount;

$pathcount >= $MAXCOUNT
    and printf "Pattern '%s' too vague, listing paths only.\n", $Pattern;

my $last = undef;
my $linesize = 0;
for my $path (sort(keys(%match)))
{
    $last = undef;
    print $path;

    if ($pathcount >= $MAXCOUNT)
    {
        print "\n";
               next;
    }

    for my $hit (@{$match{$path}})
    {
        my $inode = $$hit{inode};
        my $mtime = $$hit{mtime};
        my $image = $$hit{image}{image};
        if (defined($last) && $inode ne $last)
        {
            $linesize = 5 + length($mtime) + length($image);
            printf "\n    %s %s", $mtime, $image;
        } else {
            $linesize += length($image) + 2;
            if ($linesize > 78)
            {
                $linesize = 5 + length($mtime) + length($image);
                print "\n",
                    " " x (5 + length($mtime)),
                    $image;
            } else {
                printf ", %s", $$hit{image}{image};
            }
        }
        $last = $inode;
    }
    print "\n\n";
}

exit 0;
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
# Sort images
sub imsort_locate {
	## WARNING:  don't mess with the sort order, it is needed so that if
	## WARNING:  all images are expired the newest will be retained.
	$$a{branch} cmp $$b{branch}
	  || $$a{created} cmp $$b{created};
}
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    print STDERR <<EOUSAGE;
USAGE
	dirvish-locate vault[:branch] pattern
	
	Pattern can be any PCRE.
	
EOUSAGE
	exit 255;
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------